File: scripts/router.js
// Fixed router.js with more precise classification
const fs = require('fs');
const path = require('path');
const ROOT = path.resolve(__dirname, '..');
const RAW_DIR = path.join(ROOT, 'data', 'raw');
const ROUTING_DIR = path.join(ROOT, 'data', 'routing');

const CATEGORIES = ['multi_set', 'mixed', 'raw_questions', 'raw_analysis', 'uncategorized'];

/**
 * Improved classification:
 *
 * - Question numbers: \d+[.、] followed by letter choices (A/B/C/D) — this is a real question
 * - Analysis sections: 答案/解析/详解 appearing as section headers or in bulk,
 *   NOT just "Answer Sheet" in instructions
 * - Multi-set: 3+ Part headers
 */

function classifyFile(content) {
  // 1. Check for multi-set (3+ Part headers)
  const partMatches = content.match(/^Part\s+(I{1,3}|IV|V|1|2|3|4|5)/gmi) || [];
  if (partMatches.length >= 3) return 'multi_set';

  // 2. Check for real question numbers: digit + dot + letter choice
  //    Pattern: "1.A)" or "1. A）" or "1、A)" etc.
  const questionPattern = /\d+[.、]\s*[A-D][)）]/g;
  const questionMatches = content.match(questionPattern) || [];
  const hasQuestions = questionMatches.length >= 3; // require at least 3 questions

  // 3. Check for analysis/answer sections (not just instruction mentions)
  //    Look for: 答案与详解, 参考答案, 答案及解析, or 答案/解析 appearing as section headers
  //    Also count bulk occurrences of 解析/答案 as standalone words
  const analysisSectionPattern = /[#＃]\s*(答案|解析|详解|参考范文|Answer\s*Key|参考答案|答案与详解|答案及解析)/gi;
  const analysisSections = content.match(analysisSectionPattern) || [];

  // Also check for answer/analysis content appearing in bulk (not just instructions)
  const answerBlockPattern = /(?:^|\n)\s*(?:【?\d+】?\s*)?(?:答案|解析|详解)\s*[：:]/gm;
  const answerBlocks = content.match(answerBlockPattern) || [];

  // Check for "纯解析" style content: analysis sections + answer blocks
  const hasAnalysis = analysisSections.length >= 2 || answerBlocks.length >= 3;

  if (hasQuestions && hasAnalysis) return 'mixed';
  if (hasQuestions) return 'raw_questions';
  if (hasAnalysis) return 'raw_analysis';
  return 'uncategorized';
}

function ensureDirs() {
  for (const cat of CATEGORIES) {
    const dir = path.join(ROUTING_DIR, cat);
    if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  }
}

function safeMove(src, dest) {
  if (fs.existsSync(dest)) {
    const ext = path.extname(dest);
    const base = path.basename(dest, ext);
    let counter = 1;
    while (fs.existsSync(dest)) {
      dest = path.join(path.dirname(dest), base + '_' + counter + ext);
      counter++;
    }
  }
  fs.renameSync(src, dest);
  return dest;
}

function run(dryRun = false) {
  if (!fs.existsSync(RAW_DIR)) {
    console.error('Error: raw directory not found:', RAW_DIR);
    process.exit(1);
  }

  if (!dryRun) ensureDirs();

  const mdFiles = fs.readdirSync(RAW_DIR).filter(f => f.endsWith('.md')).sort();
  if (mdFiles.length === 0) {
    console.log('No .md files found in', RAW_DIR);
    return;
  }

  console.log('Found', mdFiles.length, '.md files in', RAW_DIR);
  console.log('Mode:', dryRun ? 'DRY RUN' : 'LIVE');
  console.log('='.repeat(60));

  const stats = {};
  for (const cat of CATEGORIES) stats[cat] = [];

  for (const filename of mdFiles) {
    const srcPath = path.join(RAW_DIR, filename);
    let content;
    try {
      content = fs.readFileSync(srcPath, 'utf-8');
    } catch (e) {
      console.log('  [ERROR] Cannot read:', filename, e.message);
      stats['uncategorized'].push(filename);
      continue;
    }

    const category = classifyFile(content);
    const destPath = path.join(ROUTING_DIR, category, filename);

    if (dryRun) {
      console.log('  [DRY]', filename, '->', category + '/');
    } else {
      const finalPath = safeMove(srcPath, destPath);
      console.log('  [MOVE]', filename, '->', category + '/' + path.basename(finalPath));
    }

    stats[category].push(filename);
  }

  console.log('='.repeat(60));
  console.log('Classification Summary:');
  console.log('-'.repeat(60));
  let total = 0;
  for (const cat of CATEGORIES) {
    const count = stats[cat].length;
    total += count;
    console.log('  ' + cat.padEnd(20) + ': ' + String(count).padStart(3) + ' files');
  }
  console.log('-'.repeat(60));
  console.log('  ' + 'TOTAL'.padEnd(20) + ': ' + String(total).padStart(3) + ' files');
  console.log('='.repeat(60));
}

const dryRun = process.argv.includes('--dry-run');
run(dryRun);

File: scripts/router.py
#!/usr/bin/env python3
"""
File Sorting Router for Exam Data Pipeline

Scans data/raw/ for .md files and classifies them into data/routing/ subdirectories
based on content analysis (not filename).

Usage:
    python scripts/router.py [--dry-run]

Categories:
    raw_questions  - Contains question numbers but no answer/analysis sections
    raw_analysis   - Contains answer/analysis sections but no question numbers
    mixed          - Contains both question numbers and answer/analysis sections
    multi_set      - Contains 3+ Part headers (multi-set exam papers)
    uncategorized  - Does not match any of the above
"""

import os
import re
import sys
import shutil
from pathlib import Path

# ===== Configuration =====

SCRIPT_DIR = Path(__file__).resolve().parent
PROJECT_ROOT = SCRIPT_DIR.parent
RAW_DIR = PROJECT_ROOT / 'data' / 'raw'
ROUTING_DIR = PROJECT_ROOT / 'data' / 'routing'

# Question pattern: number + dot/comma + letter choice (A/B/C/D)
QUESTION_RE = re.compile(r'\d+[.、]\s*[A-D][)）]', re.MULTILINE)

# Analysis section headers (not just instruction mentions)
ANALYSIS_SECTION_RE = re.compile(
    r'[#＃]\s*(答案|解析|详解|参考范文|Answer\s*Key|参考答案|答案与详解|答案及解析)',
    re.IGNORECASE | re.MULTILINE
)

# Answer blocks: numbered answers with colon
ANSWER_BLOCK_RE = re.compile(
    r'(?:^|\n)\s*(?:【?\d+】?\s*)?(?:答案|解析|详解)\s*[：:]',
    re.MULTILINE
)

# Part headers for multi-set detection
PART_RE = re.compile(
    r'Part\s+(I{1,3}|IV|V|1|2|3|4|5)',
    re.IGNORECASE | re.MULTILINE
)

CATEGORIES = ['multi_set', 'mixed', 'raw_questions', 'raw_analysis', 'uncategorized']


def classify_file(content):
    """Classify a file based on its content."""
    # 1. Multi-set: 3+ Part headers
    parts = PART_RE.findall(content)
    if len(parts) >= 3:
        return 'multi_set'

    # 2. Question numbers: require at least 3 questions
    question_matches = QUESTION_RE.findall(content)
    has_questions = len(question_matches) >= 3

    # 3. Analysis sections: require 2+ section headers or 3+ answer blocks
    analysis_sections = ANALYSIS_SECTION_RE.findall(content)
    answer_blocks = ANSWER_BLOCK_RE.findall(content)
    has_analysis = len(analysis_sections) >= 2 or len(answer_blocks) >= 3

    if has_questions and has_analysis:
        return 'mixed'
    elif has_questions:
        return 'raw_questions'
    elif has_analysis:
        return 'raw_analysis'
    else:
        return 'uncategorized'


def ensure_dirs():
    """Create routing subdirectories."""
    for cat in CATEGORIES:
        (ROUTING_DIR / cat).mkdir(parents=True, exist_ok=True)


def safe_move(src, dst):
    """Safely move a file, handling name collisions."""
    if dst.exists():
        stem = dst.stem
        suffix = dst.suffix
        counter = 1
        while dst.exists():
            dst = dst.parent / f'{stem}_{counter}{suffix}'
            counter += 1
    shutil.move(str(src), str(dst))
    return dst


def run(dry_run=False):
    """Main routing logic."""
    if not RAW_DIR.exists():
        print(f'Error: raw directory not found: {RAW_DIR}')
        sys.exit(1)

    if not dry_run:
        ensure_dirs()

    md_files = sorted(RAW_DIR.glob('*.md'))
    if not md_files:
        print(f'No .md files found in {RAW_DIR}')
        return

    print(f'Found {len(md_files)} .md files in {RAW_DIR}')
    print(f'Mode: {"DRY RUN" if dry_run else "LIVE"}')
    print('=' * 60)

    stats = {cat: [] for cat in CATEGORIES}

    for md_file in md_files:
        try:
            content = md_file.read_text(encoding='utf-8')
        except Exception as e:
            print(f'  [ERROR] Cannot read {md_file.name}: {e}')
            stats['uncategorized'].append(md_file.name)
            continue

        category = classify_file(content)
        dest = ROUTING_DIR / category / md_file.name

        if dry_run:
            print(f'  [DRY] {md_file.name} -> {category}/')
        else:
            final_path = safe_move(md_file, dest)
            print(f'  [MOVE] {md_file.name} -> {category}/{final_path.name}')

        stats[category].append(md_file.name)

    # Print summary
    print('=' * 60)
    print('Classification Summary:')
    print('-' * 60)
    total = 0
    for cat in CATEGORIES:
        count = len(stats[cat])
        total += count
        print(f'  {cat:20s}: {count:3d} files')
    print('-' * 60)
    print(f'  {"TOTAL":20s}: {total:3d} files')
    print('=' * 60)


if __name__ == '__main__':
    dry_run = '--dry-run' in sys.argv
    run(dry_run=dry_run)
